water/lib/Dialect/Wave/Transforms/LowerReadWriteOps.cpp

```diff
@@ -385,17 +406,54 @@ struct MemAccessInfo {
   int64_t vectorizedDim;
 };
 
+// Helper to get the effective type information for memory operations.
+// For wave.allocate results, uses the distributed shape for index calculations.
+// For other memory, uses the original type.
+static std::pair<ArrayRef<Attribute>, AffineMap> getEffectiveShapeInfo(
+    Value memory, wave::WaveTensorType originalType) {
+
+  // Check if this memory comes from wave.allocate
+  auto definingOp = memory.getDefiningOp();
+  if (auto allocateOp = dyn_cast_or_null<wave::AllocateOp>(definingOp)) {
+    // Use distributed shape information
+    wave::WaveExprListAttr distributedShape = allocateOp.getDistributedShape();
+    return {distributedShape.getSymbols(), distributedShape.getMap()};
+  }
+
+  // Not from wave.allocate, use original type shape
+  ArrayRef<Attribute> symbols(originalType.getShape().begin(), originalType.getShape().end());
+  return {symbols, AffineMap{}};  // Empty map means identity
+}
+
 // Common lowering for memory operations such as read/write. Checks whether
 // enough information is present on the operation and constructs the starting
 // index (populates the trailing vector argument) and eventually the mask.
 template <typename OpTy>
 static FailureOr<MemAccessInfo> createMemoryIndicesAndMask(
     ConversionPatternRewriter &rewriter, const TypeConverter *typeConverter,
-    OpTy op, wave::WaveTensorType memoryType, VectorType vectorType) {
+    OpTy op, wave::WaveTensorType memoryType, VectorType vectorType,
+    std::optional<std::pair<ArrayRef<Attribute>, AffineMap>> effectiveShapeInfo = std::nullopt) {
 
   int64_t elementsPerThread = vectorType.getNumElements();
 
-  ArrayRef<wave::WaveSymbolAttr> orderedSyms = memoryType.getShape();
+  // Use effective shape info if provided (for wave.allocate distributed shapes)
+  ArrayRef<wave::WaveSymbolAttr> orderedSyms;
+  if (effectiveShapeInfo.has_value()) {
+    ArrayRef<Attribute> symbols = effectiveShapeInfo->first;
+    SmallVector<wave::WaveSymbolAttr> convertedSyms;
+    convertedSyms.reserve(symbols.size());
+    for (Attribute attr : symbols) {
+      auto symbolAttr = dyn_cast<wave::WaveSymbolAttr>(attr);
+      if (!symbolAttr) {
+        return rewriter.notifyMatchFailure(
+            op, "expected WaveSymbolAttr in distributed shape");
+      }
+      convertedSyms.push_back(symbolAttr);
+    }
+    orderedSyms = convertedSyms;
```

Comment:
Unlike shapes in the type, distrbuted shapes may be arbitrarily complex expressions AFAIK, which may need to be handled before beyond just the list of symbols.